airstack-modules/modules/airstack/domain-name/domain-name.ts

```diff
@@ -0,0 +1,1152 @@
+  ): void {
+    const labelHash = crypto.keccak256(ByteArray.fromUTF8(name));
+    if (!labelHash.equals(label)) {
+      log.warning(
+        "Expected '{}' to hash to {}, but got {} instead. Skipping.",
+        [name, labelHash.toHex(), label.toHex()]
+      );
+      return;
+    }
+
+    if (name.indexOf(".") !== -1) {
+      log.warning("Invalid label '{}'. Skipping.", [name]);
+      return;
+    }
+    let block = getOrCreateAirBlock(chainId, blockHeight, blockHash, blockTimestamp);
+    let domain = getOrCreateAirDomain(new Domain(
+      crypto.keccak256(rootNode.concat(label)).toHex(),
+      chainId,
+      block,
+      tokenAddress,
+    ));
+
+    // tracking registration cost in domain entity  - renewal cost is not being tracked yet
+    if (fromRegistrationEvent) {
+      domain.registrationCost = cost;
+      if (paymentToken) {
+        domain.paymentToken = getOrCreateAirToken(chainId, paymentToken).id;
+      }
+      domain.lastBlock = block.id;
+    } else {
+      // name renewal event
+      // updating renewal cost in name renewed transaction entity
+      getOrCreateAirNameRenewedTransaction(
+        transactionHash,
+        chainId,
+        block,
+        domain,
+        cost,
+        paymentToken,
+        renewer!,
+        expiryTimestamp!,
+      );
+    }
+    if (domain.labelName !== name) {
+      domain.labelName = name
+      domain.name = name + '.eth'
+      domain.lastBlock = block.id;
+      // creating reverse registrar to get domainId when setting primary domain
+      if (domain.name) {
+        createReverseRegistrar(domain.name!, domain.id, block);
+      }
+    }
+    domain.save();
+    //new name registered event
+  }
+
+  /**
+   * @dev This function tracks a addr changed transaction
+   * @param chainId chain id
+   * @param logIndex event log index
+   * @param resolverAddress resolver contract address
+   * @param addr new addr
+   * @param node domain node
+   * @param blockHeight block number
+   * @param blockHash block hash
+   * @param blockTimestamp block timestamp
+   * @param transactionHash transaction hash
+   * @param tokenAddress contract address of nft token
+   */
+  export function trackAddrChangedTransaction(
+    chainId: string,
+    logIndex: BigInt,
+    resolverAddress: string,
+    addr: string,
+    node: string,
+    blockHeight: BigInt,
+    blockHash: string,
+    blockTimestamp: BigInt,
+    transactionHash: string,
+    tokenAddress: string,
+  ): void {
+    let addrAccount = getOrCreateAirAccount(chainId, addr);
+    let block = getOrCreateAirBlock(chainId, blockHeight, blockHash, blockTimestamp);
+    let domain = getOrCreateAirDomain(new Domain(node, chainId, block, tokenAddress));
+    let previousResolvedAddressId = domain.resolvedAddress;
+    let resolver = getOrCreateAirResolver(domain, chainId, resolverAddress, addr);
+    resolver.domain = domain.id;
+    resolver.save()
+
+    if (domain.resolver == resolver.id) {
```

Comment:
this is because people can set the default resolved address for a domain to any address, even if the address is not owned by them.
this check ensures that the domain resolver and resolver emitted in the resolvedAddress (Addr) changed event is same.